### PR TITLE
Add closed shadow DOM capture via CDP

### PR DESCRIPTION
**Pull request by aryanku-dev**
## Summary
- Add `expose_closed_shadow_roots()` using CDP to discover and expose closed shadow roots
- Add `_walk_nodes()` helper for CDP DOM tree traversal
- Add 6 unit tests covering tree walking, non-Chromium skip, CDP errors

Ported from percy/percy-playwright#609

## Test plan
- [ ] Unit tests pass
- [ ] Verify no regression on pages without shadow DOM

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: percy/screenshot.py [hunk 1]
```diff
@@ -96,6 +96,91 @@ def fetch_percy_dom():
     return response.text
 
 
+def _walk_nodes(node, closed_pairs):
+    """Walk CDP DOM tree to find closed shadow roots, skipping iframe boundaries."""
+    if "contentDocument" in node:
```
Comment (pranavz28) [inline percy/screenshot.py:136]: **HIGH — Walker skips ALL iframes; selenium-dotnet #436 had the same initial bug and was fixed on CE re-review.**

```python
if "contentDocument" in node:
    return
```

Same-origin iframes share the parent JS realm and the `window.__percyClosedShadowRoots` WeakMap that `PercyDOM.serialize` reads. Closed shadow roots inside same-origin iframes SHOULD be captured. Cross-origin iframes should still be skipped (different JS realm; `resolveNode` objectIds wouldn't belong to our execution context anyway).

selenium-dotnet's fix (commit `fix(snapshot): recurse closed shadow walk into same-origin iframes`):
- Same origin (`GetOrigin(documentURL) == pageOrigin`) → recurse INTO `contentDocument`.
- Cross origin → skip.
- Missing `documentURL` → defensive skip.

Suggested Python port:
```python
if "contentDocument" in node:
    cd = node["contentDocument"]
    document_url = cd.get("documentURL")
    if not document_url:
        return
    if _same_origin(document_url, page_url):
        _walk_nodes(cd, closed_pairs)
    return
```

`_same_origin` should use `urllib.parse.urlparse` to compare scheme + host + port. Add tests for all three branches (same-origin captured, cross-origin skipped, missing URL skipped) — mirroring the dotnet test suite added in that PR.

File: percy/screenshot.py [hunk 2]
```diff
@@ -96,6 +96,91 @@ def fetch_percy_dom():
...
+
+def expose_closed_shadow_roots(page):
+    """Use CDP to discover closed shadow roots and expose them to PercyDOM.serialize().
+    Closed shadow roots are inaccessible from JS (element.shadowRoot === null),
+    but CDP's DOM domain can pierce them."""
+    cdp_session = None
+    try:
+        cdp_session = page.context.new_cdp_session(page)
+
+        cdp_session.send("DOM.enable")
```
Comment (pranavz28) [inline percy/screenshot.py:161]: **HIGH — `DOM.enable` is called but `DOM.disable` is never invoked.**

selenium-dotnet #436 and selenium-ruby #32 both pair `enable`/`disable` explicitly in their try/finally and gate `disable` on whether `enable` succeeded (so a failing `enable` doesn't emit a spurious `disable`).

`cdp_session.detach()` in the finally block does mostly release the domain in practice, but the explicit pairing is the documented pattern and matches every sibling SDK that uses CDP.

Suggested fix:
```python
dom_enabled = False
try:
    cdp_session.send("DOM.enable")
    dom_enabled = True
    # ... existing walk + resolve + register ...
finally:
    if dom_enabled:
        try:
            cdp_session.send("DOM.disable")
        except Exception:
            pass
    try:
        cdp_session.detach()
    except Exception:
        pass
```

Also add a test mirroring dotnet's `DOM.disable` lifecycle suite: `DOM.disable` is sent after a successful run, sent after a mid-walk failure, NOT sent when `DOM.enable` itself raised.

File: percy/screenshot.py [hunk 3]
```diff
@@ -492,6 +577,10 @@ def percy_snapshot(page, name, **kwargs):
         # Inject the DOM serialization script
         percy_dom_script = fetch_percy_dom()
         page.evaluate(percy_dom_script)
+
+        # Expose closed shadow roots via CDP before serialization
+        expose_closed_shadow_roots(page)
```
Comment (pranavz28) [inline percy/screenshot.py:743]: **MEDIUM — `expose_closed_shadow_roots` is called once but the WeakMap is wiped on responsive reloads.**

Responsive captures reload the page at each width and re-evaluate `percy_dom_script`. The `window.__percyClosedShadowRoots` WeakMap is gone after `page.reload()`, so closed shadow roots will be missing in responsive DOMs at widths beyond the first.

selenium-ruby #32 explicitly re-primes after refresh in its responsive snapshot loop (`lib/percy.rb:549`). Search every `page.evaluate(percy_dom_script)` and `page.reload()` site in this file and precede each with `expose_closed_shadow_roots(page)`.

Add a unit test that drives the responsive capture path through `change_window_dimension_and_wait` and asserts the WeakMap is re-populated at the second width.

**Review by pranavz28 — COMMENTED** [review]
Cross-SDK review across the 6 ports of the iframe + closed-shadow-DOM rollout. This PR has the narrowest scope (closed-shadow-DOM only — no CORS iframe work), which is fine. Two HIGH findings inline + one MEDIUM around responsive-reload re-exposure.

### Verified-pass items

- Hardcoded JS in `Runtime.callFunctionOn` — no user-controlled string interpolation, low security risk.
- 7 unit tests covering tree walking, non-Chromium skip, CDP errors, detach-error suppression.
- No module-level shared state; concurrent snapshots on different pages are isolated by WeakMap key.
- `cache.py` defensive cleanup (drive-by) is safe.

### Verdict
**Request changes.** Two HIGH issues to align with selenium-dotnet #436's closed-shadow port (same-origin iframe walker + DOM.disable pairing). One MEDIUM around responsive-reload re-exposure (selenium-ruby #32 got this right; Python should mirror).

**Comment by aryanku-dev** [comment]
## 🤖 stack:pr-review — Automated review (BrowserStack AI harness)

**Summary:** Adds closed-shadow-DOM capture via CDP for Chromium (`expose_closed_shadow_roots` + `_walk_nodes`/`_get_origin`/`_same_origin`), recursing into same-origin iframe content documents while skipping cross-origin; re-primes the `__percyClosedShadowRoots` WeakMap after responsive reloads; hardens `Cache.cleanup_cache` against missing keys / concurrent mutation.

| Priority | Category | Check | Status |
|---|---|---|---|
| High | Security | No hardcoded secrets | Pass |
| High | Security | Input validation | Pass — `_get_origin` parses defensively, returns "" on bad URLs; origin compared scheme+netloc |
| High | Security | AuthZ / IDOR / SQLi | N/A |
| High | Correctness | Logic & edge cases | Pass — same-origin recursion matches dotnet canonical; cross/unknown-origin skipped; empty `closed_pairs` early-return |
| High | Correctness | Explicit error handling | Pass — failures debug-logged; `# pragma: no cover` only on best-effort DOM.disable/detach cleanup |
| High | Correctness | No race conditions | Pass — `cleanup_cache` iterates a `list(...)` copy + `.get()` — fixes dict-mutation race |
| Medium | Testing | New code / error paths tested | Pass — `TestClosedShadowDOM` (12) + cache orphan + responsive re-prime; DOM.disable covered in all 3 branches |
| Medium | Testing | Existing tests pass | Pass — 95 tests OK, **100% branch coverage** |
| Medium | Performance | No unbounded fetch | Pass — DOM walked once; CDP bounded by closed-root count |
| Medium | Quality | Patterns / focused | Pass — faithful port of `percy-selenium-dotnet` canonical |
| Low | Quality | Names / comments / deps | Pass — stdlib `urllib.parse` only |

**Findings:** no Fail. Non-blocking note: dotnet wraps each `(host, shadow)` pair in its own try/except; the Python version relies on the outer try (first failing pair stops the loop) — still safe/logged/non-fatal and matches the canonical JS, left as-is to keep the diff minimal.

**Verification:** pylint **10.00/10**; full unittest suite **95 OK**; coverage **100%**. The three original reviewer threads (same-origin iframe recursion, DOM.disable pairing, WeakMap re-prime) all hold.

**Commits:** the fixes are at `e7e8626` (already on `PER-7292`); nothing further needed this pass.

### Overall: ✅ **Pass** — Approve. No human-decision blockers.

<sub>Posted by the BrowserStack `stack:pr-review` harness.</sub>

**Comment by aryanku-dev** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/percy-playwright-python/pull/136  •  **Head:** 3b9bdaf  •  **Reviewers:** stack:code-reviewer

## Summary
Adds closed shadow DOM capture via CDP (`expose_closed_shadow_roots` / `_walk_nodes`) and honors `data-percy-ignore` + `ignoreIframeSelectors` on the CORS-iframe path, plus a resilience fix to `cleanup_cache`. The CDP flow pairs `DOM.enable`/`DOM.disable` behind a `dom_enabled` flag, guards each host resolution individually, and always detaches in `finally`. The closed-shadow WeakMap is re-primed after responsive reloads. SDK-local options are stripped from both serialize args and the CLI POST body. I independently inspected all three of pranavz28's review concerns: two are already fixed; one is a defensible, documented design tradeoff matching the canonical SDKs (non-gating).

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No secrets/credentials | Pass | No tokens/keys in diff. |
| High | Security | Input validation / injection | Pass | `page.evaluate` passes `frame.url`/selectors as bound args (not interpolated); serialize args via `json.dumps`. |
| High | Security | Safe deserialization / eval | Pass | No `eval`/`pickle`/`yaml.load`; CDP JSON handled defensively. |
| High | Security | AuthN/AuthZ | N/A | No auth surface touched. |
| High | Correctness | No crashes / unhandled exceptions | Pass | CDP path wrapped try/except/finally; per-host guard; `_frame_ignore_flags` falls back to all-False on error. |
| High | Correctness | Resource cleanup | Pass | `DOM.disable` gated on `dom_enabled`; `cdp_session.detach()` always attempted in `finally`. |
| High | Correctness | Iteration safety | Pass | `cleanup_cache` iterates `list(cls.CACHE.items())`, avoiding mutation-during-iteration. |
| Medium | Testing | New code has tests | Pass | Tests cover walker, exposure, per-host failure, enable-failed vs mid-walk disable paths, selector resolution, e2e skip/capture. |
| Medium | Testing | Error/edge cases tested | Pass | Missing `TIMEOUT_KEY` skipped; empty/None/non-list selectors normalized; detached frames handled. |
| Medium | Testing | Existing tests pass | Pass | CI green (Test 3.8/3.9/3.10, Lint, Analyze, Semgrep). |
| Medium | Performance | No needless work / N+1 | Pass | Ignore check runs before expensive inject/serialize; `fetch_percy_dom` still `lru_cache`d. |
| Medium | Performance | Heavy work backgrounded | N/A | Synchronous capture path. |
| Medium | Quality | Follows patterns | Pass | Mirrors JS SDK semantics; shared `_get_origin`/`_same_origin` helpers; backwards-compatible kwargs. |
| Medium | Quality | Focused change | Pass | Scoped to closed-shadow + ignore-selector handling plus the cache fix. |
| Low | Quality | Names / no dead code | Pass | Clear helper names; `page_origin` retained intentionally (documented compat). |
| Low | Quality | Comments explain why | Pass | Docstrings explain realm/WeakMap invariants and compat shims. |
| Low | Quality | No unnecessary deps | Pass | None added. |

## Findings
No independently-confirmed High/Critical findings.

Non-blocking observation: `_walk_nodes` still threads `page_origin` purely for signature/test compatibility while ignoring it (documented). Harmless; the dead parameter could be removed in a future cleanup once tests are updated.

### Raised by other reviewers (not independently confirmed)
**pranavz28 (HIGH): "Walker skips ALL iframes; should recurse same-origin."** Inspected `_walk_nodes`: it returns on any node with a `contentDocument`, including same-origin iframes. The author rebuts this in the docstring — a child-frame document (even same-origin) is a separate JS realm, so `DOM.resolveNode` of a host inside it yields an `objectId` in a realm where `window.__percyClosedShadowRoots` is undefined, causing `Runtime.callFunctionOn` to throw; same-origin grants cross-frame DOM access but not a shared JS execution context. The canonical SDKs (percy-playwright JS, percy-puppeteer, percy-playwright-java) all `return` on `contentDocument` too. I do not independently confirm this as a definite High bug — it is a defensible design tradeoff (cross-frame closed shadow roots remain unsupported, degrading gracefully rather than throwing). Non-gating; worth a maintainer decision, not a code-review block.

pranavz28's other two concerns were inspected and found **already fixed**: (b) `DOM.enable`/`DOM.disable` pairing via the `dom_enabled` flag (with tests), and (c) `expose_closed_shadow_roots` re-primed after the responsive reload (with a test). Not gating.

---

**Verdict: PASS**

**Review by pranavz28 — APPROVED** [review]
